spec/services/idv/steps/send_link_step_spec.rb

```diff
@@ -0,0 +1,84 @@
+require 'rails_helper'
+
+describe Idv::Steps::SendLinkStep do
+  let(:user) { build(:user) }
+
+  let(:service_provider) do
+    create(
+      :service_provider,
+      issuer: 'http://sp.example.com',
+      app_id: '123',
+    )
+  end
+
+  let(:request) do
+    double(
+      'request',
+      remote_ip: Faker::Internet.ip_v4_address,
+      headers: { 'X-Amzn-Trace-Id' => amzn_trace_id },
+    )
+  end
+
+  let(:params) do
+    ActionController::Parameters.new(
+      {
+        doc_auth: { phone: '(201) 555-1212' },
+      },
+    )
+  end
+
+  let(:controller) do
+    instance_double(
+      'controller',
+      session: { sp: { issuer: service_provider.issuer } },
+      params: params,
+      current_user: user,
+      analytics: FakeAnalytics.new,
+      url_options: {},
+      request: request,
+    )
+  end
+
+  let(:amzn_trace_id) { SecureRandom.uuid }
+
+  let(:pii_from_doc) do
+    {
+      ssn: '123-45-6789',
+      first_name: 'bob',
+    }
+  end
+
+  let(:flow) do
+    Idv::Flows::DocAuthFlow.new(controller, {}, 'idv/doc_auth').tap do |flow|
+      flow.flow_session = { pii_from_doc: pii_from_doc }
+    end
+  end
+
+  let(:irs_attempts_api_tracker) do
+    IrsAttemptsApiTrackingHelper::FakeAttemptsTracker.new
+  end
+
+  subject(:step) do
+    Idv::Steps::SendLinkStep.new(flow)
+  end
+
+  describe 'the return value from #call' do
+    let(:response) { step.call }
+
+    before do
+      allow(controller).to receive(:irs_attempts_api_tracker).
+        and_return(irs_attempts_api_tracker)
+    end
+
+    it 'includes the telephony response' do
+      expect(response.extra[:telephony_response]).to eq(
```

Comment:
can we add specs that makes sure this gets logged to the analytics instance? not that it just creates the keys?

something like:

```
expect(controller.analytics).to have_logged_event('the event name', message_id: 'fake-message-id', request_id: 'fake-message-request-id')
```

Reply:
Since this is testing `step.call` it doesn't actually call the analytics service. The analytics service gets called in the controller [here](https://github.com/18F/identity-idp/blob/1d4ba7381a60cd12c6655d89889ef260e2ec6dc2/app/services/flow/flow_state_machine.rb#L26). That is well after `SendLinkStep#call` is called.

As a result, if we want to make sure this is passed to the analytics service we will need to do one of two things:

a.) Refactor so `SendLinkStep#call` makes a special call to the analytics service with the telephony results
b.) Add a test for this to the doc auth controller.

I don't think option B would be worth the effort. Option A may be worth something. But if we want the telehpony response in the send link event I think the appropriate way to test that is to make sure it is part of the result and then depend on the tests for the flow state machine to ensure that result gets logged.

Reply:
ah good points!